Feat:[PKD-003][PKD-006][PKD-020]Pokemon Information Page: Moves & Fix…

Enrickzz opened the pull request:
… Get Pokemon Type

- details_pokemon_handler update to get pokemon moves
- Get Pokemon Type on PostFrameCallback
- base_stats ui updates
- pokemon details page UI updates (colors)
- added pokemon_moves_tab widget (UI and Implementation)
- added universal_functions.dart and color constants in constants.dart

https://user-images.githubusercontent.com/61441278/181159752-eeaa6df5-b9e8-49ce-b9b4-490567ba08c0.mp4

Enrickzz committed:
Feat:[PKD-003][PKD-006][PKD-020]Pokemon Information Page: Moves & Fix Get Pokemon Type

- details_pokemon_handler update to get pokemon moves
- Get Pokemon Type on PostFrameCallback
- base_stats ui updates
- pokemon details page UI updates (colors)
- added pokemon_moves_tab widget (UI and Implementation)
- added universal_functions.dart and color constants in constants.dart

dbenedictos commented on lib/feature/home_page/widgets/pokemon_tile.dart:
```diff
@@ -7,6 +7,7 @@ import 'package:pokedex_async_redux/feature/pokemon_details_page.dart/pokemon_de
+import 'package:pokedex_async_redux/utilities/universal_functions.dart';
 
 class PokemonTile extends StatefulWidget {
   const PokemonTile({
```
trailing comma

dbenedictos commented on lib/feature/home_page/widgets/pokemon_tile.dart:
```diff
@@ -22,25 +23,30 @@ class _PokemonTileState extends State<PokemonTile> {
-                  url: e['type']['url'],
-                ))).toList(),
-          );
+    if (mounted) {
```
put the mounted check inside the post frame callback. Reason is that if this holds true, and then, for some reason, the widget suddenly has been unmounted after the mounted check, everything inside the callback will still be called.

dbenedictos replied:
![Screenshot 2022-07-27 at 1 38 08 PM](https://user-images.githubusercontent.com/47196695/181169569-b037d9e3-8d17-48d9-bb5a-37e33e5aacf3.png)
for example. Still encountered the error 😛

dbenedictos commented on lib/feature/pokemon_details_page.dart/pokemon_details_page.dart:
```diff
@@ -58,13 +62,13 @@ class PokemonDetails extends StatelessWidget {
                                     margin: EdgeInsets.fromLTRB(0, 0, 5, 0),
                                     decoration: BoxDecoration(
-                                      color: Colors.grey, // TODO: Make dynamic later
+                                      color: getPokemonColor('typeDetailsPage'), // TODO: Make dynamic later
```
strings file

dbenedictos commented on lib/feature/pokemon_details_page.dart/pokemon_details_page.dart:
```diff
@@ -78,13 +82,13 @@ class PokemonDetails extends StatelessWidget {
                                     child: Container(
                                       decoration: BoxDecoration(
-                                        color: Colors.grey, // change color later
+                                        color: getPokemonColor('typeDetailsPage'),
```
string files

dbenedictos commented on lib/feature/pokemon_details_page.dart/pokemon_details_page.dart:
```diff
@@ -156,10 +163,12 @@ class PokemonDetails extends StatelessWidget {
+                                      Container(color: Colors.white),
+                                      MovesTab(
+                                          moves: pokemon.moves,
+                                          color: getPokemonColor(pokemon.types?.first.type?.name ?? '')),
```
trailing comma

dbenedictos commented on lib/utilities/universal_functions.dart:
```diff
@@ -0,0 +1,55 @@
+import 'package:flutter/material.dart';
+import 'package:pokedex_async_redux/utilities/constants.dart';
+
+String capitalizeFirstLetter(String name) {
```
instead of a function, make an extension instead.

dbenedictos replied:
when you're done, put the extension inside the extensions.dart file inside your utilities folder

dbenedictos commented on lib/utilities/universal_functions.dart:
```diff
@@ -0,0 +1,55 @@
+      return kSteel;
+    case 'water':
+      return kWater;
+    case 'typeDetailsPage':
```
what is `typeDetailsPage`?

Enrickzz replied:
to get the color for this widget
![image](https://user-images.githubusercontent.com/61441278/181167682-bd331445-dbf0-4093-aae0-cd75f5844aa5.png)

dbenedictos replied:
do you use the white colour? If not make this the default instead

dbenedictos commented on lib/utilities/universal_functions.dart:
```diff
@@ -0,0 +1,55 @@
+  }
+}
+
+Color getPokemonColor(String type) {
```
I think this can be an extension too.

dbenedictos commented on lib/feature/pokemon_details_page.dart/pokemon_details_page.dart:
```diff
@@ -78,13 +82,13 @@ class PokemonDetails extends StatelessWidget {
                                         : true,
                                     child: Container(
```
Make a separate widget for this, so you can just reuse the widget wherever.

dbenedictos replied:
line 80: 'pokemon.types?.last.type?.name == pokemon.types?.first.type?.name' already returns a boolean, so why make is a ternary?

Enrickzz committed:
PR Feedback

- Fix: get pokemon type : setstate if mounted only
- renamed universal_functions to extensions
- converted functions to extensions
- Wrapped moves tab to SingleChildScrollView
- updated function calls to fit extensions
- added widget for Type
- added const color for TypeWidget background
- formatting (trailing commas)

dbenedictos commented on lib/utilities/extensions.dart:
```diff
@@ -0,0 +1,48 @@
+import 'package:flutter/material.dart';
+import 'package:pokedex_async_redux/utilities/constants.dart';
+
+extension StringExt on String {
```
noice

dbenedictos commented on lib/feature/home_page/widgets/pokemon_tile.dart:
```diff
@@ -7,26 +7,25 @@ import 'package:pokedex_async_redux/feature/pokemon_details_page.dart/pokemon_de
-        var result = jsonDecode(value.body);
-        List SubTypeMap = result['types'];
-        setState(() {
+    WidgetsBinding.instance.addPostFrameCallback((_) {
```
postframe((){
   if (mounted){
       doSomething();
   }
})

dbenedictos replied:
async

dbenedictos commented on lib/feature/home_page/widgets/pokemon_tile.dart:
```diff
@@ -36,14 +35,23 @@ class _PokemonTileState extends State<PokemonTile> {
     super.initState();
   }
 
+  @override
```
no need

dbenedictos commented on lib/utilities/constants.dart:
```diff
@@ -1,6 +1,29 @@
+const Color kSteel = Color(0xff7c8e8a);
+const Color kFairy = Color(0xfffda1c2);
+
+const Color bgTypeDetailsPage = Color.fromARGB(99, 255, 255, 255);
```
`typeDetailsPageBackgroundColor`

dbenedictos commented on lib/feature/pokemon_details_page.dart/pokemon_details_page.dart:
```diff
@@ -14,13 +17,13 @@ class PokemonDetails extends StatelessWidget {
           SliverAppBar(
             leading: BackButton(color: Colors.black),
-            backgroundColor: Colors.white, // TODO: Make dynamic later
+            backgroundColor: pokemon.types?.first.type?.name?.getPokemonColor ?? ''.getPokemonColor,
```
for readability
`(pokemon.types?.first.type?.name?.getPokemonColor ?? '').getPokemonColor`
and do the same for the rest

dbenedictos commented on lib/utilities/extensions.dart:
```diff
@@ -0,0 +1,48 @@
+import 'package:pokedex_async_redux/utilities/constants.dart';
+
+extension StringExt on String {
+  String get capitalize => '${this[0].toUpperCase()}${substring(1)}';
```
some strings are nullable when you call `captilize` also add null checks here

dbenedictos replied:
ahh no, never mind

dbenedictos commented on lib/feature/widgets/type_container_widget.dart:
```diff
@@ -0,0 +1,32 @@
+
+class TypeWidget extends StatelessWidget {
+  const TypeWidget({
+    required this.type,
```
trailing comma

dbenedictos commented on lib/feature/pokemon_details_page.dart/widgets/pokemon_moves_tab.dart:
```diff
@@ -0,0 +1,38 @@
+        children: moves
+                ?.map((e) => Padding(
+                      padding: EdgeInsets.all(2),
+                      child: Container(
```
use your TypeContainerWidget here

dbenedictos commented on lib/feature/widgets/type_container_widget.dart:
```diff
@@ -0,0 +1,32 @@
+import 'package:pokedex_async_redux/utilities/constants.dart';
+import 'package:pokedex_async_redux/utilities/extensions.dart';
+
+class TypeWidget extends StatelessWidget {
```
match your widget name to your file name. or vice-versa

Enrickzz committed:
PR Feedbacks

- if mounted revision
- converted type_container to pill_container (more generic)
- renamed constants

dbenedictos commented on lib/feature/home_page/widgets/pokemon_tile.dart:
```diff
@@ -7,26 +7,25 @@ import 'package:pokedex_async_redux/feature/pokemon_details_page.dart/pokemon_de
-        List SubTypeMap = result['types'];
-        setState(() {
+    WidgetsBinding.instance.addPostFrameCallback((_) {
+      http.get(Uri.tryParse('${widget.thisPokemon.url}') ?? Uri()).then((value) {
```
await

Enrickzz committed:
PR Feedback

- remove dispose() (not needed)

dbenedictos commented on lib/feature/home_page/widgets/pokemon_tile.dart:
```diff
@@ -36,14 +35,22 @@ class _PokemonTileState extends State<PokemonTile> {
-      } else {
-        print('error');
-      }
+          if (mounted) {
```
put this if statement after the outside the `then` method

dbenedictos commented on lib/feature/home_page/widgets/pokemon_tile.dart:
```diff
@@ -36,14 +35,22 @@ class _PokemonTileState extends State<PokemonTile> {
   }
 
+  @override
+  void dispose() {
```
no need for the dispose override. because you're not doing anything inside of it

dbenedictos reviewed (changes requested):
Last set of comments

Enrickzz committed:
PR Feedback

- add async await keys
- moved if mounted after .then
